### PR TITLE
Feat/add vote function  #16

### PR DESCRIPTION
**Pull request by Horlakz**
#### Feature(s)

- basic CRUD operations for votes
- installed express-async-handler for handling errors better
- made actions field in models required

### CODE REVIEW

File: src/controllers/voteController.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  try {
+    // check if action is valid
+    if (body.action !== "VOTE_UP" && body.action !== "VOTE_DOWN") {
+      throw new Error(
+        "Please provide a valid action, either VOTE_UP or VOTE_DOWN"
+      );
+    }
+
+    // check if user has already voted
+    const checkVote = await Vote.findOne({
```
Comment (struckchure) [inline src/controllers/voteController.js:22]: use `Vote.exists(...)` instead

File: src/controllers/voteController.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+// POST /vote/
+const addVote = asyncHandler(async (req, res) => {
+  const { body } = req;
+
+  const add_vote = new Vote(body);
+
+  try {
+    // check if action is valid
+    if (body.action !== "VOTE_UP" && body.action !== "VOTE_DOWN") {
```
Comment (struckchure) [inline src/controllers/voteController.js:15]: the enum ensures the option is either `VOTE_UP` or `VOTE_DOWN`, no need to validate it

File: src/controllers/voteController.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      );
+    }
+
+    // check if user has already voted
+    const checkVote = await Vote.findOne({
+      user_id: body.user_id,
+      content_id: body.content_id,
+    });
+
+    if (checkVote) {
```
Comment (struckchure) [inline src/controllers/voteController.js:27]: make it so if they make the request twice it negates the previous voting entry, like if it was `VOTE_UP`, change it to `VOTE_DOWN`, vice versa

File: src/routes/forumRoutes.js [hunk 4]
```diff
@@ -38,5 +45,10 @@ router
   .put(updateReply)
   .delete(deleteReply);
 
+// vote routes
+router.post("/vote/", addVote);
+router.route("/vote/:content_id").get(getVotes).put(updateVote);
+router.delete("/vote/:content_id/:vote_id", removeVote);
```
Comment (struckchure) [inline src/routes/forumRoutes.js:51]: pass `content_id` and `vote_id` as query parameter
`?content_id=some-id&vote_id=some-vote-id`

**Review by struckchure — CHANGES_REQUESTED** [review]

File: src/controllers/voteController.js (hunk 2 above)
Comment (Horlakz) [reply src/controllers/voteController.js:15]: - [ x ]

Comment (Horlakz) [reply src/controllers/voteController.js:15]: Done

**Pull request merged**